### PR TITLE
ENH: np.DataSource for transparent compression and remote files

### PR DESCRIPTION
**Pull request by ksunden**
- [x] Implementation (those that use genfromtxt needed no changes)
- [x] Document from a users perspective (add to docstrings, mention in quickstart, maybe a few other places in human generated rst files)
- [x] Document from developers perspective (Edit page on writing a from method with explicit instructions on how to get it to work.)
    - [x] Any call to `open` should be `ds = np.DataSource(None); f = ds.open(fname, 'rt');...;f.close()`
- [x] Write tests for behaviour, particularly remote files. Already have made files compressed, may desire to compress all ascii text files, which would then want at least a few tests of the uncompressed file for good measure.
- [ ] ~Decide whether to include some remote files in datasets (could go so far as making only remote), could also make those files persistent and add code to look for previous downloads.~ Not this PR, but keep in mind for the future, perhaps the near future with Kohler2017 data?, Regardless, that engineering is not part of this already pretty hefty PR.

### CODE REVIEW

**Comment by ksunden** [comment]
Perhaps we should use bz2 instead of gz for one of the solis datasets, It would ensure that both work

**Comment by ksunden** [comment]
Kind of reminds me of #365 in a way. not the same, but perhaps some of the same ideas.

If we ever wish to revisit the idea of a cache for datasets, (or possibly in some future iteration of `wt.datasets`) DataSource may provide a way of doing it fairly simply. May consider taking in an optional kwarg in from methods for a DataSource object (or path to create one)... 

It's an idea, but not a fully thought out one at this point.

**Comment by ksunden** [comment]
Test failures are because I'm using pre-release tidy_headers here

File: WrightTools/collection/_directory.py [hunk 1]
```diff
@@ -66,26 +67,26 @@ def from_directory(filepath, from_methods, *, name=None, parent=None, verbose=Tr
...
                         parent=parent,
                         verbose=verbose,
                     )
                 break
         else:
-            if os.path.isdir(os.path.join(path, fname)):
+            if (path / fname).is_dir():
                 if verbose:
-                    print("Creating Collection at", posixpath.join(parent.name, fname))
+                    print("Creating Collection at", pathlib.PosixPath(parent.name) / fname)
```
Comment (ksunden) [inline WrightTools/collection/_directory.py]: perhaps should be PurePosixPath here...

**Comment by ksunden** [comment]
I apologize to the review for the volume of change in this PR, but it is mostly repetitive changes, so that helps.

File: setup.py [hunk 2]
```diff
@@ -44,7 +44,7 @@ def read(fname):
         "python-dateutil",
         "pytz",
         "scipy",
-        "tidy_headers>=0.3.0",
+        "tidy_headers>=1.0.0",
```
Comment (darienmorrow) [inline setup.py]: This will conflict with #813

**Pull request merged**